rpmio/macro.c

```diff
@@ -1306,8 +1306,22 @@ static void doFoo(rpmMacroBuf mb, rpmMacroEntry me, ARGV_t argv, size_t *parsed)
 	    b++;
     } else if (rstreq("dirname", me->name)) {
 	buf = xstrdup(argv[1]);
-	if ((b = strrchr(buf, '/')) != NULL)
-	    *b = '\0';
+	if (!strcmp(buf, "")) {
+	    buf = rstrcat(&buf, ".");
+	} else if (strcmp(buf, ".") && strcmp(buf, "..") && strcmp(buf, "/")) {
+	    /* strip trailing / */
+	    size_t l = strlen(buf);
+	    if (buf[l-1] == '/')
+		buf[l-1] = '\0';
+	    if ((b = strrchr(buf, '/')) != NULL)
+		if (b == buf)
+		    /* keep root dir */
+		    buf[1] = '\0';
+		else
+		    *b = '\0';
+	    else
+		strcpy(buf, ".");
+	}
```

Comment:
Um, why not call dirname() instead?

Reply:
Good question. Why didn't the code used that in the first place...?

Reply:
It's a good question. You have to go pretty far back (commit 94f5fbeec363059ca9eb986ee294ffcdf20bcb04) and come back none the wiser :smile: